Dockerfile

```diff
@@ -1,61 +1,153 @@
     && echo "cores = $CORES" >> /etc/nix/nix.conf \
-    && nix profile install nixpkgs#cachix \
+    && nix profile install nixpkgs#cachix nixpkgs#git-lfs \
     && cachix use kernel-builder
-
 WORKDIR /kernelcode
 COPY . /etc/kernel-builder/
-
 ENV MAX_JOBS=${MAX_JOBS}
 ENV CORES=${CORES}
-
 RUN mkdir -p /etc/kernelcode && \
-    cat <<'EOF' > /etc/kernelcode/entry.sh
+    cat <<'EOF' > /etc/kernelcode/cli.sh
 #!/bin/sh
-echo "Building Torch Extension Bundle"
+set -e
 
-# Check if kernelcode is a git repo and get hash if possible
-if [ -d "/kernelcode/.git" ]; then
-  # Mark git as safe to allow commands
-  git config --global --add safe.directory /kernelcode
+# Default values
+BUILD_URL=""
+DEV_SHELL=0
+HELP=0
 
-  # Try to get git revision
-  REV=$(git rev-parse --short=8 HEAD)
-  
-  # Check if working directory is dirty
-  if [ -n "$(git status --porcelain 2)" ]; then
-    REV="${REV}-dirty"
+# CLI usage function
+function show_usage {
+  echo "Kernel Builder CLI"
+  echo ""
+  echo "Usage: docker run [docker-options] kernel-builder:dev [command] [options]"
+  echo ""
+  echo "Commands:"
+  echo "  build               Build the kernel extension (default if no command specified)"
+  echo "  dev                 Start a development shell"
+  echo "  fetch [URL]         Clone and build from a Git URL"
+  echo "  help                Show this help message"
+  echo ""
+  echo "Options:"
+  echo "  --jobs, -j NUMBER   Set maximum number of parallel jobs (default: $MAX_JOBS)"
+  echo "  --cores, -c NUMBER  Set number of cores per job (default: $CORES)"
+  echo ""
+  echo "Examples:"
+  echo "  docker run -v \$(pwd):/kernelcode kernel-builder:dev build"
+  echo "  docker run -it -v \$(pwd):/kernelcode kernel-builder:dev dev"
+  echo "  docker run kernel-builder:dev fetch https://huggingface.co/user/repo.git"
+}
+
+# Function to build the extension
+function build_extension {
+  echo "Building Torch Extension Bundle"
+  # Check if kernelcode is a git repo and get hash if possible
+  if [ -d "/kernelcode/.git" ]; then
+    # Mark git as safe to allow commands
+    git config --global --add safe.directory /kernelcode
+    # Try to get git revision
+    REV=$(git rev-parse --short=8 HEAD)
+    
+    # Check if working directory is dirty
+    if [ -n "$(git status --porcelain 2)" ]; then
+      REV="${REV}-dirty"
+    fi
+  else
+    # Generate random material if not a git repo
+    REV=$(dd if=/dev/urandom status=none bs=1 count=10 2>/dev/null | base32 | tr '[:upper:]' '[:lower:]' | head -c 10)
   fi
-else
-  # Generate random material if not a git repo
-  REV=$(dd if=/dev/urandom status=none bs=1 count=10 2>/dev/null | base32 | tr '[:upper:]' '[:lower:]' | head -c 10)
-fi
+  echo "Building with rev $REV"
+  nix build \
+      --impure \
+      --max-jobs $MAX_JOBS \
+      -j $CORES \
+      --expr "with import /etc/kernel-builder; lib.x86_64-linux.buildTorchExtensionBundle { path = /kernelcode; rev = \"$REV\"; }" \
+      -L
```

Comment:
Maybe we should assume that there is a `flake.nix` (and do a build without `--impure`)? Or alternatively, create a `flake.nix` for the user if it does not exist and then do a pure build?

Reply:
great idea, added in the latest commits